File: diameterOfBinaryTree.java

public class diameterOfBinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    static int maxDiameter = 0;

    public static int diameterOfBinaryTree(TreeNode root) {
        calculateHeight(root);
        return maxDiameter;
    }
    
    public static int calculateHeight(TreeNode root) {
        if (root == null) {
            return 0;
        } else {
            int leftHeight = calculateHeight(root.left);
            int rightHeight = calculateHeight(root.right);
            maxDiameter = Math.max(maxDiameter, leftHeight + rightHeight);

            return 1 + Math.max(leftHeight, rightHeight);
        }
    }
}

File: binaryTree/diameterOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def height(root, result):
            if not root:
                return 0
            
            l, r = height(root.left, result), height(root.right, result)

            diameter = l + r
            if diameter > result[0]:
                result[0] = diameter

            return max(l, r) + 1
        
        result = [0]
        height(root, result)
        return result[0]